conftest.py

```diff
@@ -0,0 +1,226 @@
+        seed = random.randint(0, 2**31-1)
+    else:
+        seed = int(module_seed_str)
+        logging.warning('*** module-level seed is set: '
+                        'all tests running deterministically ***')
+    print('Setting module np/mx/python random seeds, '
+          'use MXNET_MODULE_SEED={} to reproduce.'.format(seed))
+
+    random.seed(seed)
+    try:
+        import numpy as np
+        import mxnet as mx
+        np.random.seed(seed)
+        mx.random.seed(seed)
+    except:
+        print('Unable to import numpy/mxnet. Skipping conftest.')
+
+    # The MXNET_TEST_SEED environment variable will override MXNET_MODULE_SEED for tests with
+    #  the 'with_seed()' decoration.  Inform the user of this once here at the module level.
+    if os.getenv('MXNET_TEST_SEED') is not None:
+        logging.warning('*** test-level seed set: all "@with_seed()" '
+                        'tests run deterministically ***')
+
+
+@pytest.hookimpl(tryfirst=True, hookwrapper=True)
+def pytest_runtest_makereport(item, call):
+    """Make test outcome available to fixture.
+
+    https://docs.pytest.org/en/latest/example/simple.html#making-test-result-information-available-in-fixtures
+    """
+    # execute all other hooks to obtain the report object
+    outcome = yield
+    rep = outcome.get_result()
+
+    # set a report attribute for each phase of a call, which can
+    # be "setup", "call", "teardown"
+    setattr(item, "rep_" + rep.when, rep)
+
+
+@pytest.fixture(scope='function', autouse=True)
+def function_scope_seed(request):
+    """A function scope fixture that manages rng seeds.
+
+    This fixture automatically initializes the python, numpy and mxnet random
+    number generators randomly on every test run.
+
+    def test_ok_with_random_data():
+        ...
+
+    To fix the seed used for a test case mark the test function with the
+    desired seed:
+
+    @pytest.mark.seed(1)
+    def test_not_ok_with_random_data():
+        '''This testcase actually works.'''
+        assert 17 == random.randint(0, 100)
+
+    When a test fails, the fixture outputs the seed used. The user can then set
+    the environment variable MXNET_TEST_SEED to the value reported, then rerun
+    the test with:
+
+        pytest --verbose -s <test_module_name.py> -k <failing_test>
+
+    To run a test repeatedly, install pytest-repeat and add the --count argument:
+
+        pip install pytest-repeat
+        pytest --verbose -s <test_module_name.py> -k <failing_test> --count 1000
+
+    """
+
+    seed = request.node.get_closest_marker('seed')
+    env_seed_str = os.getenv('MXNET_TEST_SEED')
+
+    if seed is not None:
+        seed = seed.args[0]
+        assert isinstance(seed, int)
+    elif env_seed_str is not None:
+        seed = int(env_seed_str)
+    else:
+        seed = random.randint(0, 2**31-1)
+
+    random.seed(seed)
+    try:
+        import numpy as np
+        import mxnet as mx
+        post_test_state = np.random.get_state()
```

Comment:
what about the mx.random state? Are we unable to save it?

Reply:
yes, we are unable to save it yet as we don't expose the states in the frontend yet.